### DISCUSSION TITLE
Is hyper Server utilize threads?

### DISCUSSION THREAD

**Discussion by BratSinot** [hyperium/hyper#3016] [Q&A] [answered]
Greetings!

Before tokio, hyper utilized threads: https://docs.rs/hyper/0.9.5/hyper/server/index.html
> A Server is created to listen on port, parse HTTP requests, and hand them off to a Handler. By default, the Server will listen across multiple threads, but that can be configured to a single thread if preferred.

But what about now? Should I manually spawn a task for every incoming connection to utilize the whole CPU?

**Comment by seanmonstar** [comment, marked as answer]
If you're using `hyper::Server`, it will spawn a [task](https://docs.rs/tokio/latest/tokio/task/index.html) for each connection.

If you're using `hyper::server::conn::Http` combined with your own socket listener, then you could spawn each HTTP connection into a task.

The runtime (such as Tokio) will use multiplex the tasks across the amount of threads you've configured it to use (Tokio will use the same amount of threads as CPUs if you didn't specify otherwise).

**Reply by BratSinot** [reply]
And if I use `Server` like this?
```rust
let server = Server::builder(incoming).serve(app);
```

**Reply by seanmonstar** [reply]
That's what I meant, if you're using `hyper::Server`, it will spawn a task per connection.

**Reply by BratSinot** [reply]
Great! Thanks!
I tried to find something in sources, but have no success =)

**Reply by BratSinot** [reply]
> That's what I meant, if you're using `hyper::Server`, it will spawn a task per connection.

Okey, then I miss something. For example there will no futures connections until wait complete:
```
curl http://localhost:3000
curl http://localhost:3000
```
P.S. By this wait I emulate TLS handshake then try connect to TLS but by http protocol.
```rust
use hyper::{
    server::{accept, Server},
    service::{make_service_fn, service_fn},
    Body, Request, Response,
};
use std::{
    convert::Infallible,
    net::SocketAddr,
    sync::atomic::{AtomicUsize, Ordering},
    time::Duration,
};
use tokio::net::TcpListener;
use tokio_stream::wrappers::TcpListenerStream;
use tokio_stream::StreamExt;

#[tokio::main]
async fn main() {
    static RCX: AtomicUsize = AtomicUsize::new(0);

    let addr = SocketAddr::from(([127, 0, 0, 1], 3000));

    let listener = TcpListener::bind(addr).await.unwrap();
    let stream = TcpListenerStream::new(listener).then(|stream| async move {
        if RCX.fetch_add(1, Ordering::SeqCst) % 2 == 0 {
            println!("wait");
            // there can be TLS handshake
            tokio::time::sleep(Duration::from_secs(10)).await;
        } else {
            println!("no wait");
        }

        stream
    });
    let incoming = accept::from_stream(stream);

    Server::builder(incoming)
        .serve(make_service_fn(|_conn| async {
            Ok::<_, Infallible>(service_fn(|req: Request<Body>| async move {
                Ok::<_, Infallible>(Response::new(Body::from(format!("{req:?}"))))
            }))
        }))
        .await
        .unwrap()
}
```

**Reply by seanmonstar** [reply]
Yes, there is a shortcoming in v0.14.x, that the accept handshake part is done in the same task. Once a connection is accepted, it's spawned into a new task. You can either make the `accept` part spawn handshake tasks, or use `hyper::server::conn::Http` and manage tasks yourself.

**Reply by BratSinot** [reply]
> Yes, there is a shortcoming in v0.14.x, that the accept handshake part is done in the same task.

Any issue for this?
